pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go

```diff
@@ -0,0 +1,232 @@
+package nodeshutdown
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/godbus/dbus/v5"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/clock"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/kubelet/eviction"
+	"k8s.io/kubernetes/pkg/kubelet/nodeshutdown/systemd"
+	kubelettypes "k8s.io/kubernetes/pkg/kubelet/types"
+)
+
+const (
+	nodeShutdownReason  = "Shutdown"
+	nodeShutdownMessage = "Node is shutting, evicting pods"
+)
+
+var getSystemDbus = systemDbus
+
+type dbusInhibiter interface {
+	CurrentInhibitDelay() (time.Duration, error)
+	InhibitShutdown() (systemd.InhibitLock, error)
+	ReleaseInhibitLock(lock systemd.InhibitLock) error
+	ReloadLogindConf() error
+	MonitorShutdown() (<-chan bool, error)
+	OverrideInhibitDelay(inhibitDelayMax time.Duration) error
+}
+
+// Manager has functions that can be used to interact with the Node Shutdown Manager.
+type Manager struct {
+	shutdownGracePeriodRequested    time.Duration
+	shutdownGracePeriodCriticalPods time.Duration
+
+	getPods eviction.ActivePodsFunc
+	killPod eviction.KillPodFunc
+
+	dbusCon               dbusInhibiter
+	nodeShuttingDownNow   bool
+	nodeShuttingDownMutex sync.Mutex
+	inhibitLock           systemd.InhibitLock
+
+	clock clock.Clock
+}
+
+func systemDbus() (dbusInhibiter, error) {
+	bus, err := dbus.SystemBus()
+	if err != nil {
+		return nil, err
+	}
+	return &systemd.DBusCon{SystemBus: bus}, nil
+}
+
+// NewManager returns a new node shutdown manager.
+func NewManager(getPodsFunc eviction.ActivePodsFunc, killPodFunc eviction.KillPodFunc, shutdownGracePeriodRequested, shutdownGracePeriodCriticalPods time.Duration) *Manager {
+	return &Manager{
+		getPods:                         getPodsFunc,
+		killPod:                         killPodFunc,
+		shutdownGracePeriodRequested:    shutdownGracePeriodRequested,
+		shutdownGracePeriodCriticalPods: shutdownGracePeriodCriticalPods,
+		clock:                           clock.RealClock{},
+	}
+}
+
+// Start starts the node shutdown manager and will start watching the node for shutdown events.
+func (m *Manager) Start() error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.GracefulNodeShutdown) {
+		return nil
+	}
+	if m.shutdownGracePeriodRequested == 0 {
+		return nil
+	}
+
+	systemBus, err := getSystemDbus()
+	if err != nil {
+		return err
+	}
+	m.dbusCon = systemBus
+
+	currentInhibitDelay, err := m.dbusCon.CurrentInhibitDelay()
+	if err != nil {
+		return err
+	}
+
+	// If the logind's InhibitDelayMaxUSec as configured in (logind.conf) is less than shutdownGracePeriodRequested, attempt to update the value to shutdownGracePeriodRequested.
+	if m.shutdownGracePeriodRequested > currentInhibitDelay {
+		err := m.dbusCon.OverrideInhibitDelay(m.shutdownGracePeriodRequested)
+		if err != nil {
+			return err
```

Comment:
will error be descriptive enough? Should it be wrapped here?

Reply:
good point, wrapped the error here with a more descriptive message.